drivers: dac: Add DAC support for Renesas RA devices

khoa-nguyen-18 opened the pull request:
This PR is to support DAC driver for the Renesas boards:
- RA8: ek_ra8m1, ek_ra8d1, mck_ra8t1
- RA6: ek_ra6m5, ek_ra6m4, ek_ra6m3, ek_ra6m2, ek_ra6m1, ek_ra6e2, fpb_ra6e2, fpb_ra6e1
- RA4: ek_ra4m3, ek_ra4m2, ek_ra4e2
- RA2: ek_ra2a1

khoa-nguyen-18 committed:
manifest: Update hal_renesas for RA DAC support

Update hal_renesas to support DAC on RA

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 committed:
drivers: dac: Initial DAC driver support for Renesas RA

Initial DAC driver support for Renesas RA

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 committed:
dts: arm: renesas: ra: Initial support DAC for Renesas RA

Add DAC node to support DAC for ra8x1, ra6-cm4, ra6-cm33,
ra4-cm33, ra2xx soc

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 committed:
boards: renesas: Initial support DAC driver for Renesas RA

Add support DAC driver for ek_ra8m1, ek_ra8d1, mck_ra8t1,
ek_ra6m5, ek_ra6m4, ek_ra6m3, ek_ra6m2, ek_ra6m1, ek_ra6e2,
fpb_ra6e2, fpb_ra6e1, ek_ra4m3, ek_ra4m2, ek_ra4e2, ek_ra2a1

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 committed:
tests: drivers: dac: Add support to test DAC on RA devices

Add support ek_ra8m1, ek_ra8d1, mck_ra8t1, ek_ra6m5,
ek_ra6m4, ek_ra6m3, ek_ra6m2, ek_ra6m1, ek_ra6e2,
fpb_ra6e2, fpb_ra6e1, ek_ra4m3, ek_ra4m2, ek_ra4e2,
ek_ra2a1 to test:
- zephyr/tests/drivers/dac/dac_api
- zephyr/tests/drivers/dac/dac_loopback (except ek_ra2a1)

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 committed:
samples: drivers: dac: Add support for RA to test sample DAC

Add overlay to support RA to test sample DAC

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 committed:
dts: arm: renesas: correct the interrupt number for RA SOC

Currently the interrupt number of the ADC node is duplicated
with other node on several RA soc. This commit aim to
resolve this issue.

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

thaoluonguw commented:
@khoa-nguyen-18 : Please rebase to solve conflict and update for commit ID of hal_renesas. Thank you.

khoa-nguyen-18 commented:
> @khoa-nguyen-18 : Please rebase to solve conflict and update for commit ID of hal_renesas. Thank you.

I have rebased and updated revision of hal_renesas, thank you

martinjaeger commented on samples/drivers/dac/boards/ek_ra2a1.overlay:
```diff
@@ -0,0 +1,19 @@
+	};
+};
+
+&dac {
+	dac0: dac@0 {
+		status = "okay";
+	};
+};
```
This was already done on board-level, so we don't need to repeat it here. Same for all other overlays in this sample.

```suggestion
```

martinjaeger commented on boards/renesas/ek_ra2a1/ek_ra2a1.dts:
```diff
@@ -109,3 +109,11 @@
 	status = "okay";
 };
+
+&dac {
+	dac@0 {
+		pinctrl-0 = <&dac0_default>;
+		pinctrl-names = "default";
+		status = "okay";
+	};
+};
```
Why not like this? (also for all other boards)

```suggestion
&dac0 {
	pinctrl-0 = <&dac0_default>;
	pinctrl-names = "default";
	status = "okay";
};
```

khoa-nguyen-18 replied:
I have updated it, including for other boards as well. Thank you

martinjaeger reviewed (changes requested):
Some initial comments from my side. Will do a proper review of the driver code later.

Why did you change the ADC accuracy test to use a DAC and not the reference voltage anymore? And could that change be moved into a dedicated PR, as it doesn't seem necessary for this PR and it's more related to ADC and not DAC.

khoa-nguyen-18 commented on drivers/dac/Kconfig.renesas_ra:
```diff
@@ -0,0 +1,44 @@
+	     - n: Do not synchronize DAC12 with ADC (unit 1) operation
+	     (disable interference reduction between D/A and A/D conversion).
+	     - y: Synchronize DAC12 with ADC (unit 1) operation
+	     (enable interference reduction between D/A and A/D conversion).
```
thank you, I have updated it

khoa-nguyen-18 commented on drivers/dac/dac_renesas_ra.c:
```diff
@@ -0,0 +1,163 @@
+		}
+	}
+
+#if BSP_FEATURE_DAC_HAS_OUTPUT_AMPLIFIER
```
After considering, I have moved these HW features to the device tree to make it clearer for users. Thank you

khoa-nguyen-18 commented on drivers/dac/dac_renesas_ra.c:
```diff
@@ -0,0 +1,163 @@
+
+#if BSP_FEATURE_DAC_HAS_OUTPUT_AMPLIFIER
+	config_extend->output_amplifier_enabled = channel_cfg->buffered;
+#elif BSP_FEATURE_DAC_HAS_CHARGEPUMP
```
same as above, thank you

khoa-nguyen-18 commented:
> Why did you change the ADC accuracy test to use a DAC and not the reference voltage anymore? 

I think changing the reference source of ADC in the test app `adc_accuracy_test` will help verify both the correctness of the DAC and the ADC, rather than just the ADC.

> And could that change be moved into a dedicated PR, as it doesn't seem necessary for this PR and it's more related to ADC and not DAC.

Yes, I will move it to another PR after this DAC gets merged. I have updated it. Thank you.

martinjaeger commented on drivers/dac/dac_renesas_ra.c:
```diff
@@ -0,0 +1,170 @@
+		.f_config =                                                                        \
+			{                                                                          \
+				.channel = DT_INST_REG_ADDR(idx),                                  \
+				DAC_DA_AD_SYNCHRONIZE_PARAMETER,                                   \
```
Could this not be simplified?

```suggestion
				.ad_da_synchronized = IS_ENABLED(CONFIG_DAC_RENESAS_RA_DA_AD_SYNCHRONIZE), \
```

khoa-nguyen-18 replied:
Thanks for your suggestion, I have updated it.

martinjaeger commented on drivers/dac/Kconfig.renesas_ra:
```diff
@@ -0,0 +1,44 @@
+if DAC_RENESAS_RA
+config DAC_RENESAS_RA_DA_AD_SYNCHRONIZE
+	bool "D/A A/D Synchronous Conversion"
+	default n
```
`n` should not be explicitly set.

```suggestion
```

khoa-nguyen-18 replied:
Thank you, I have removed it

martinjaeger commented on drivers/dac/dac_renesas_ra.c:
```diff
@@ -0,0 +1,170 @@
+		.enable_charge_pump = true,                                                        \
+		.output_amplifier_enabled = true,                                                  \
+		.internal_output_enabled = false,                                                  \
+		DAC_DAVREFCR_PARAMETER,                                                            \
```
I prefer not to hide the name of the struct member behind a `#define`. This should also allow to reduce the `#ifdef`s above. You will need to assign `0` to `DAC_RENESAS_RA_DAVREFCR` in a final `#else` block.

```suggestion
		.ref_volt_sel = DAC_RENESAS_RA_DAVREFCR,                                                            \
```

khoa-nguyen-18 replied:
I agree with your suggestion. I have updated it. Thank you.

khoa-nguyen-18 commented on dts/bindings/dac/renesas,ra-dac-global.yaml:
```diff
@@ -0,0 +1,25 @@
+
+  renesas,dac-has-davrefcr:
+    type: boolean
+    description: True if the SoC supports the D/A VREF configuration in the DAC HWIP
```
Thanks for your suggestion, I have updated it

khoa-nguyen-18 commented on dts/bindings/dac/renesas,ra-dac.yaml:
```diff
@@ -0,0 +1,15 @@
+    const: 1
+
+io-channel-cells:
+  - input
```
It was my mistake. Thank you for pointing it out. I have updated it.

khoa-nguyen-18 commented on drivers/dac/dac_renesas_ra.c:
```diff
@@ -0,0 +1,170 @@
+		LOG_ERR("Requires MOCO clock to support the buffer feature");
+		return -ENOTSUP;
+	}
+#endif
```
Yes, `MOCO clock` also needs to be enabled for the charge pump to work. I have updated it to make the `LOG_ERR` clearer.

martinjaeger replied:
Ok, but where and when would the `MOCO clock` be turned on under normal circumstances? I see no calls to the Zephyr `clock_control` driver in the init function of this driver.

khoa-nguyen-18 replied:
Currently, we configure and enable the `MOCO clock` in hal_renesas, and its activation depends on the `status` of the MOCO node in the device tree.

martinjaeger replied:
Ok, clock control should probably be moved to the Zephyr repo in the future, but it should work for now. Thanks for the explanation.
